File: Samples/Verbum/how-to-automatically-determine-a-document-format-based-on-document-content-t343926-14.2.9-/CS/WindowsFormsApplication1/RichEditControlExtensions.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using DevExpress.Utils.Zip;
using DevExpress.XtraRichEdit;
using DevExpress.XtraRichEdit.Services;

namespace RichEditControlExtensions {
    public static class RichEditLoadDocumentHelper {
        const byte P = (byte)'P';
        const byte K = (byte)'K';

        static readonly byte[] docSignature = { 0xd0, 0xcf, 0x11, 0xe0, 0xa1, 0xb1, 0x1a, 0xe1 };


        public static void RegisterCusomOpenFileCommand(this RichEditControl richEdit) {
            CustomRichEditCommandFactoryService commandFactory = new CustomRichEditCommandFactoryService(richEdit, richEdit.GetService<IRichEditCommandFactoryService>());
            richEdit.RemoveService(typeof(IRichEditCommandFactoryService));
            richEdit.AddService(typeof(IRichEditCommandFactoryService), commandFactory);
        }

        public static void LoadDocumentEx(this IRichEditControl richEdit, Stream stream) {
            DevExpress.XtraRichEdit.DocumentFormat currentFormat = AutoDetermineDocumentFormat(stream);
            stream.Seek(0, SeekOrigin.Begin);
            richEdit.InnerControl.LoadDocument(stream, currentFormat);
        }

        public static void LoadDocumentEx(this IRichEditControl richEdit, string fileName) {
            using(System.IO.FileStream fs = new FileStream(fileName, System.IO.FileMode.Open)) {
                richEdit.LoadDocumentEx(fs);
            }
        }

        public static DevExpress.XtraRichEdit.DocumentFormat AutoDetermineDocumentFormat(Stream stream) {
            byte[] buffer = new byte[8];

            if(stream.Length < 8) return DevExpress.XtraRichEdit.DocumentFormat.PlainText;
            stream.Read(buffer, 0, buffer.Length);

            if(buffer.SequenceEqual(docSignature))
                return DevExpress.XtraRichEdit.DocumentFormat.Doc;

            else if(buffer[0] == P && buffer[1] == K)
                return CheckZippedFileContent(stream);

            string fileBeginning = Encoding.ASCII.GetString(buffer);
            if(fileBeginning.Contains("{\\rtf")) return DevExpress.XtraRichEdit.DocumentFormat.Rtf;
            if(fileBeginning.Contains("<!")) return DevExpress.XtraRichEdit.DocumentFormat.Html;

            return DevExpress.XtraRichEdit.DocumentFormat.PlainText;
        }

        static DevExpress.XtraRichEdit.DocumentFormat CheckZippedFileContent(Stream stream) {
            stream.Seek(0, SeekOrigin.Begin);
            InternalZipFileCollection files = InternalZipArchive.Open(stream);
            foreach(InternalZipFile entry in files) {
                if(entry.FileName.Contains(".rels") || entry.FileName.Contains("document.xml")) return DevExpress.XtraRichEdit.DocumentFormat.OpenXml;
                if(entry.FileName.Contains("content.xml")) return DevExpress.XtraRichEdit.DocumentFormat.OpenDocument;
            }
            return DevExpress.XtraRichEdit.DocumentFormat.PlainText;
        }
    }
}

File: Samples/Verbum/how-to-automatically-determine-a-document-format-based-on-document-content-t343926-14.2.9-/VB/WindowsFormsApplication1/RichEditControlExtensions.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text
Imports DevExpress.Utils.Zip
Imports DevExpress.XtraRichEdit
Imports DevExpress.XtraRichEdit.Services

Namespace RichEditControlExtensions
	Public Module RichEditLoadDocumentHelper
		Private Const P As Byte = CByte(AscW("P"c))
		Private Const K As Byte = CByte(AscW("K"c))

        Private ReadOnly docSignature() As Byte = {&HD0, &HCF, &H11, &HE0, &HA1, &HB1, &H1A, &HE1}

        <System.Runtime.CompilerServices.Extension> _
        Public Sub RegisterCusomOpenFileCommand(ByVal richEdit As RichEditControl)
            Dim commandFactory As New CustomRichEditCommandFactoryService(richEdit, richEdit.GetService(Of IRichEditCommandFactoryService)())
            richEdit.RemoveService(GetType(IRichEditCommandFactoryService))
            richEdit.AddService(GetType(IRichEditCommandFactoryService), commandFactory)
        End Sub

        <System.Runtime.CompilerServices.Extension> _
        Public Sub LoadDocumentEx(ByVal richEdit As IRichEditControl, ByVal stream As Stream)
            Dim currentFormat As DevExpress.XtraRichEdit.DocumentFormat = AutoDetermineDocumentFormat(stream)
            stream.Seek(0, SeekOrigin.Begin)
            richEdit.InnerControl.LoadDocument(stream, currentFormat)
        End Sub

        <System.Runtime.CompilerServices.Extension> _
        Public Sub LoadDocumentEx(ByVal richEdit As IRichEditControl, ByVal fileName As String)
            Using fs As System.IO.FileStream = New FileStream(fileName, System.IO.FileMode.Open)
                richEdit.LoadDocumentEx(fs)
            End Using
        End Sub

        Public Function AutoDetermineDocumentFormat(ByVal stream As Stream) As DevExpress.XtraRichEdit.DocumentFormat
            Dim buffer(7) As Byte

            If stream.Length < 8 Then
                Return DevExpress.XtraRichEdit.DocumentFormat.PlainText
            End If
            stream.Read(buffer, 0, buffer.Length)

            If buffer.SequenceEqual(docSignature) Then
                Return DevExpress.XtraRichEdit.DocumentFormat.Doc

            ElseIf buffer(0) = P AndAlso buffer(1) = K Then
                Return CheckZippedFileContent(stream)
            End If

            Dim fileBeginning As String = Encoding.ASCII.GetString(buffer)
            If fileBeginning.Contains("{\rtf") Then
                Return DevExpress.XtraRichEdit.DocumentFormat.Rtf
            End If
            If fileBeginning.Contains("<!") Then
                Return DevExpress.XtraRichEdit.DocumentFormat.Html
            End If

            Return DevExpress.XtraRichEdit.DocumentFormat.PlainText
        End Function

        Private Function CheckZippedFileContent(ByVal stream As Stream) As DevExpress.XtraRichEdit.DocumentFormat
            stream.Seek(0, SeekOrigin.Begin)
            Dim files As InternalZipFileCollection = InternalZipArchive.Open(stream)
            For Each entry As InternalZipFile In files
                If entry.FileName.Contains(".rels") OrElse entry.FileName.Contains("document.xml") Then
                    Return DevExpress.XtraRichEdit.DocumentFormat.OpenXml
                End If
                If entry.FileName.Contains("content.xml") Then
                    Return DevExpress.XtraRichEdit.DocumentFormat.OpenDocument
                End If
            Next entry
            Return DevExpress.XtraRichEdit.DocumentFormat.PlainText
        End Function
	End Module
End Namespace
